### PR TITLE
refactor: make the phi2 example can be directly load the model from hf without convert needed

### PR DESCRIPTION
**Pull request by mzbac**
Just try to make the phi2 example compatible with the phi2 model structure on HF, so people can directly use it without needing to convert. Additionally, given the small size of the model, it will be beneficial if people could directly load the full precision on the model without converting it first. 
Also, currently the open source LLM community prefers distributing fine-tuned models with lora weight merged, so keep the model structure consistent could avoid confusion for those distributing mlx fine-tuned models later.

But please feel free to dismiss this PR if it doesn't fit the purpose.

### CODE REVIEW

File: llms/phi2/README.md [hunk 1]
```diff
@@ -7,63 +7,52 @@ GPT-4 outputs and clean web text.
...
 
+### Convert new models 
 
-## Generate
+You can convert (change the data type or quantize) models using the
+`convert.py` script. This script takes a Hugging Face repo as input and outputs
+a model directory (which you can optionally also upload to Hugging Face).
 
-To generate text with the default prompt:
+For example, to make 4-bit quantized a model, run:
```
Comment (awni) [inline llms/phi2/README.md:40]: Nit `to make a`

**Review by awni — APPROVED** [review]
I think this is great, we should definitely have it this way. Avoiding the required pre-conversion is a nice feature!

**Comment by awni** [comment]
One thing I'm wondering about is if we should attempt to merge with `hf_llm` so that it can load and generate with Phi2 models as well? What do you think @mzbac ?

We'd probably would need to check the `config["model_type"]` field when constructing the model. From what I can see the main differences are the fused qkv matmul and layer norm instead of RMS norm. Not sure if it makes to have a whole separate model file or to try and make those configurable. (I am a bit scared of configuration condition bloat which can be hard to maintain).

  Either way this PR is a step in that direction so I'm going to merge it!

**Comment by mzbac** [comment]
> One thing I'm wondering about is if we should attempt to merge with `hf_llm` so that it can load and generate with Phi2 models as well? What do you think @mzbac ?
> 
> We'd probably would need to check the `config["model_type"]` field when constructing the model. From what I can see the main differences are the fused qkv matmul and layer norm instead of RMS norm. Not sure if it makes to have a whole separate model file or to try and make those configurable. (I am a bit scared of configuration condition bloat which can be hard to maintain).
> 
> Either way this PR is a step in that direction so I'm going to merge it!

It would be great if we could merge all those hf llm models into the `hf_llm`. From what I can see in the transformer source code, it seems like they have separate models for each model and use architectures to identify the model class when instantiating them. may be we can follow the similar approach?https://github.com/huggingface/transformers/blob/main/src/transformers/models/auto/auto_factory.py#L392

**Comment by awni** [comment]
> From what I can see in the transformer source code, it seems like they have separate models for each model and use architectures to identify the model class when instantiating them. may be we can follow the similar approach?

From a simplicity standpoint I think that's the right call. There will probably be a lot of code dup. To keep it organized we could have a `models` subdir and a file for each model type

**Comment by mzbac** [comment]
> > From what I can see in the transformer source code, it seems like they have separate models for each model and use architectures to identify the model class when instantiating them. may be we can follow the similar approach?
> 
> From a simplicity standpoint I think that's the right call. There will probably be a lot of code dup. To keep it organized we could have a `models` subdir and a file for each model type

Hi @awni, I just quickly tried out the approach in one of my repositories. It looks alright and not as repetitive as I originally thought. Let me know what you think. here is the repo: https://github.com/mzbac/mlx-llm/tree/refactor/llm.

**Comment by awni** [comment]
> Hi @awni,I quickly tried out the approach in one of my repositories. It looks alright and not as repetitive as I originally thought. The flow for adding a new model would be to simply add the model args and model class, then update the mapping in utils.py. I feel that the developer experience is kind of smooth.

I love that!  I would say to keep the example relatively flat and understandable, it would be better to have your models in the directory `models/llama.py` and just put everything llama specific file (e.g. including the llama specific model args). That makes even more straightforward to add a new model (1 model is 1 file).  (We still trying to strike a good balance between simplicity and functionality in our examples and sometimes they can be at tensions with one another 😄 )

Wdyt? If that sounds good, would be great to have you PR it?

**Comment by mzbac** [comment]
@awni  Yeah, sounds good to me. The reason I put it in the folder was to provide some flexibility for future extension, but I'm happy to go with one file per model for now. We can always separate it later if needed. I'll start drafting something, hopefully we can get it done soon and enable us to make `hf_llm` a dedicated package.

**Pull request merged**